repo2docker/buildpacks/pipfile/__init__.py

```diff
@@ -65,51 +68,72 @@ def python_version(self):
             if os.path.exists(nb_requirements_file):
                 assemble_scripts.append(
                     (
                         "${NB_USER}",
-                        # want the $NB_PYHTON_PREFIX environment variable, not for
-                        # Python's string formatting to try and replace this
-                        '${{NB_PYTHON_PREFIX}}/bin/pip install --no-cache-dir -r "{}"'.format(
+                        'pip install --no-cache-dir -r "{}"'.format(
                             nb_requirements_file
                         ),
                     )
                 )
 
-        # install Pipfile.lock or fallback to installing Pipfile
-        pipenv = "${KERNEL_PYTHON_PREFIX}/bin/pipenv"
-        python = "${KERNEL_PYTHON_PREFIX}/bin/python"
         pipfile = self.binder_path("Pipfile")
         pipfile_lock = self.binder_path("Pipfile.lock")
+
+        # A Pipfile(.lock) can contain relative references, so we need to be
+        # mindful about where we invoke pipenv as that will dictate where .`
+        # referes to.
+        #     [packages]
+        #     my_package_example = {path=".", editable=true}
         working_directory = self.binder_dir or "."
+
+        # install pipenv to install dependencies within Pipfile.lock or Pipfile
         assemble_scripts.append(("${NB_USER}", "pip install pipenv"))
-        # if Pipfile.lock isn't found, Pipfile is used to create one
-        if not os.path.exists(pipfile_lock):
-            assemble_scripts.append(
-                (
-                    "${NB_USER}",
-                    "(cd {} && {} lock --python {})".format(
-                        working_directory, pipenv, python
-                    ),
-                )
-            )
-        # install Pipfile.lock
+
+        # NOTES:
+        # - Without prioritizing the PATH to KERNEL_PYTHON_PREFIX over
+        #   NB_SERVER_PYTHON_PREFIX, 'pipenv' draws the wrong conclusion about
+        #   what Python environment is the '--system' environment.
+        # - The --system flag allows us to avoid wrapping ourself in yet
+        #   another virtual environment that we also then need to enter.
+        #   This flag is only available within the `install` subcommand of
+        #   `pipenv`.
+        # - The `--skip-lock` will not run the `lock` subcommand again as
+        #   part of the `install` command. This allows a preexisting .lock
+        #   file to remain intact and be used directly. This allows us to
+        #   prioritize usage of .lock files that makes sense for
+        #   reproducibility.
+        # - The `--ignore-pipfile` requires a .lock file to be around as if
+        #   there isn't, no other option remain.
         assemble_scripts.append(
             (
                 "${NB_USER}",
-                "(cd {} && {} install --ignore-pipfile --deploy --system --dev --python {})".format(
-                    working_directory, pipenv, python
+                """(\\
```

Comment:
What does a `\\` do?

Reply:
A `\\` renders to `\` in the end within the Dockerfile, allowing content to be combined to one single line. It allowed me to split up the content that renders to the Dockerfile into multiple lines.

Reply:
I removed the initial one being alone.

Reply:
I also added a comment about what those were about!